libcudacxx/include/cuda/__iterator/transform_output_iterator.h

```diff
@@ -0,0 +1,500 @@
+             && _CUDA_VSTD::is_nothrow_assignable_v<_CUDA_VSTD::iter_reference_t<const _Iter>,
+                                                    _CUDA_VSTD::invoke_result_t<const _Fn&, _Arg>>)
+  {
+    *__iter_ = _CUDA_VSTD::invoke(__func_, _CUDA_VSTD::forward<_Arg>(__arg));
+    return *this;
+  }
+};
+
+//! \addtogroup iterators
+//! \{
+
+//!! \addtogroup fancyiterator Fancy Iterators
+//!  \ingroup iterators
+//!  \{
+
+//! @brief \p transform_output_iterator is an iterator which represents a pointer into a range of values after
+//! transformation by a function. This iterator is useful for creating a range filled with the result of applying an
+//! operation to another range without either explicitly storing it in memory, or explicitly executing the
+//! transformation. Using
+//! \p transform_output_iterator facilitates kernel fusion by deferring the execution of a transformation until the
+//! value is needed while saving both memory capacity and bandwidth.
+//!
+//! The following code snippet demonstrates how to create a \p transform_output_iterator which represents the result of
+//! \c sqrtf applied to the contents of a \p thrust::device_vector.
+//!
+//! @code
+//! #include <cuda/iterator>
+//! #include <thrust/device_vector.h>
+//!
+//! struct square_root
+//! {
+//!   __host__ __device__
+//!   float operator()(float x) const
+//!   {
+//!     return sqrtf(x);
+//!   }
+//! };
+//!
+//! int main()
+//! {
+//!   thrust::device_vector<float> v{1.0f, 4.0f, 9.0f, 16.0f};
+//!
+//!   using FloatIterator = thrust::device_vector<float>::iterator;
+//!
+//!   cuda::transform_output_iterator iter(v.begin(), square_root{});
+//!
+//!   *iter;   // returns 1.0f
+//!   iter[0]; // returns 1.0f;
+//!   iter[1]; // returns 2.0f;
+//!   iter[2]; // returns 3.0f;
+//!   iter[3]; // returns 4.0f;
+//!
+//!   // iter[4] is an out-of-bounds error
+//! }
+//! @endcode
+//!
+//! This next example demonstrates how to use a \p transform_output_iterator with the \p thrust::reduce function to
+//! compute the sum of squares of a sequence. We will create temporary \p transform_output_iterators utilising class
+//! template argument deduction avoid explicitly specifying their type:
+//!
+//! @code
+//! #include <cuda/iterator>
+//! #include <thrust/device_vector.h>
+//! #include <thrust/reduce.h>
+//! #include <iostream>
+//!
+//! struct square
+//! {
+//!   __host__ __device__
+//!   float operator()(float x) const
+//!   {
+//!     return x * x;
+//!   }
+//! };
+//!
+//! int main()
+//! {
+//!   // initialize a device array
+//!   thrust::device_vector<float> v(4);
+//!   v[0] = 1.0f;
+//!   v[1] = 2.0f;
+//!   v[2] = 3.0f;
+//!   thrust::device_vector<float> v{1.0f, 2.0f, 3.0f, 4.0f};
+//!   thrust::reduce(cuda::transform_output_iterator{v.begin(), square{}},
+//!                  cuda::transform_output_iterator{v.end(),   square{}});
+//!
+//!   std::cout << "sum of squares: " << sum_of_squares << std::endl;
+//!   return 0;
+//! }
```

Comment:
Important: I think this example would work for a `transform_iterator`, but not a `transform_output_iterator`. Please drop this example.